[Sensor] SensorSuite in each SceneNode

vauduong opened the pull request:
## Motivation and Context
Due to new functionality from https://github.com/facebookresearch/habitat-sim/pull/1064, there needs to be more functionality to keep track of sensors.

In this PR:
- Change the sensors_ in SensorSuite from std::map<std::string, Sensor::ptr> to std::map<std::string, reference_wrapper<Sensor>>
- Make SensorSuite inherited from Magnum::SceneGraph::AbstractFeature3D
- Add raw pointers nodeSensorSuite and subtreeSensorSuite in SceneNode. These features will be released upon SceneNode's deletion.
- In the Sensor constructor, add this sensor to its node's nodeSensorSuite and subtreeSensorSuite. Traverse all the way up to the root node, add this sensor to the subtreeSensorSuite of every ancestor node. Remove the Sensor from  subtreeSensorSuite of every ancestor node in the destructor.
- Overload setParent so that ancestor subtrees are updated 
- Write bindings for getter functions and SensorFactory methods
- Write robust unit tests for creating sensors, deleting sensors, and setting parents

![Copy of SensorCameraWeeklyUpdates 8 pptx](https://user-images.githubusercontent.com/15386768/110880706-ac2fe700-82a4-11eb-9b35-0753b09d210a.png)


## How Has This Been Tested

Build and test, write additional unit tests

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

Skylion007 commented on src/esp/scene/SceneNode.cpp:
```diff
@@ -16,13 +17,17 @@ SceneNode::SceneNode(SceneNode& parent)
   setId(parent.getId());
   setCachedTransformations(Mn::SceneGraph::CachedTransformation::Absolute);
   absoluteTransformation_ = absoluteTransformation();
+  nodeSensorSuite_ = new esp::sensor::SensorSuite(*this);
```
Don't these need corresponding destructors?

bigbike replied:
@Skylion007 : No, they do not.
`new esp::sensor::SensorSuite(*this)` will create this feature in memory, and attach it to the node (`*this`), which means the node owns this feature (`sensorSuite`). When the node gets deleted, it will first release the attached features. 
the raw pointer `nodeSensorSuite_` is to provide a shortcut to locate this feature, so that no need to traverse the entire feature list of the node.

bigbike replied:
@vivian: Would you consider putting the above reply into the comments so people will not be confused in the future? thanks.

vauduong commented on src/esp/sensor/SensorFactory.cpp:
```diff
@@ -1,22 +1,26 @@
+                                  const sensor::SensorSetup& sensorSetup) {
   for (const SensorSpec::ptr& spec : sensorSetup) {
     scene::SceneNode& sensorNode = node.createChild();
     // VisualSensor Setup
     if (spec->isVisualSensorSpec()) {
       if (spec->sensorSubType == SensorSubType::Orthographic ||
           spec->sensorSubType == SensorSubType::Pinhole) {
-        sensorSuite.add(CameraSensor::create(
-            sensorNode, std::dynamic_pointer_cast<CameraSensorSpec>(spec)));
+        sensor::CameraSensor::create(
+            sensorNode,
+            std::dynamic_pointer_cast<sensor::CameraSensorSpec>(spec));
       }
```
I'm having some trouble with initializing sensors here. Currently, the sensor is initialized and a reference is added to all the relevant node and subtree sensorSuites, as expected. However, as soon as it is done initializing and the for loop moves onto the next SensorSpec, the sensor is destructed immediately. I'm not sure why this is happening because the reference to the sensor should still exist as it was added to a map, right? Why is this happening and how can I get the Sensor to persist?

bigbike replied:
> I'm not sure why this is happening because the reference to the sensor should still exist as it was added to a map, right? Why is this happening and how can I get the Sensor to persist?

When the smart pointer gets destroyed, the dtor of Sensor class is called, which means the sensor you just added to the map gets removed (you just coded it up by yourself in this PR).

bigbike commented on src/esp/sensor/SensorFactory.cpp:
```diff
@@ -1,22 +1,26 @@
-            sensorNode, std::dynamic_pointer_cast<CameraSensorSpec>(spec)));
+        sensor::CameraSensor::create(
+            sensorNode,
+            std::dynamic_pointer_cast<sensor::CameraSensorSpec>(spec));
```
This is a good one. @mosra 
You will have to register this feature in that node. Try one of the following:
```
sensorNode.addFeature<sensor::CameraSensor>(std::dynamic_pointer_cast<sensor::CameraSensorSpec>(spec));
```
Or this one:
```
       new sensor::CameraSensor{
           sensorNode,
           std::dynamic_pointer_cast<sensor::CameraSensorSpec>(spec)};
```

vauduong replied:
Thanks!

bigbike commented on src/esp/scene/SceneNode.h:
```diff
@@ -41,6 +46,7 @@ class SceneNode : public MagnumObject,
   // terminate node (e.g., "MagnumScene" defined in SceneGraph) as its ancestor
   SceneNode() = delete;
   SceneNode(SceneNode& parent);
+  ~SceneNode() { LOG(INFO) << "Deconstructing sceneNode"; };
```
~SceneNode() = default;

Otherwise it would be too verbose.

bigbike commented on src/esp/sensor/Sensor.cpp:
```diff
@@ -51,9 +50,31 @@ Sensor::Sensor(scene::SceneNode& node, SensorSpec::ptr spec)
+  node().getNodeSensorSuite().remove(*this);
+  node().getSubtreeSensorSuite().remove(*this);
+  // Traverse up to root node and remove sensor from every subtreeSensorSuite
+  auto parent = node().parent();
```
change `parent` to `currentNode` to increase the readability.

bigbike commented on src/esp/sensor/Sensor.cpp:
```diff
@@ -67,18 +88,28 @@ void Sensor::setTransformationFromSpec() {
-Sensor::ptr SensorSuite::get(const std::string& uuid) const {
-  return (sensors_.at(uuid));
+void SensorSuite::remove(sensor::Sensor& sensor) {
+  const std::string uuid = sensor.specification()->uuid;
```
you do not need to construct this local `uuid`.

bigbike commented on src/esp/sensor/Sensor.cpp:
```diff
@@ -67,18 +88,28 @@ void Sensor::setTransformationFromSpec() {
-  return (sensors_.at(uuid));
+void SensorSuite::remove(sensor::Sensor& sensor) {
+  const std::string uuid = sensor.specification()->uuid;
+  std::map<std::string, std::reference_wrapper<sensor::Sensor>>::iterator it =
```
you can do key-based `erase` which means you do not have to find it first.

bigbike commented on src/esp/sensor/Sensor.cpp:
```diff
@@ -67,18 +88,28 @@ void Sensor::setTransformationFromSpec() {
 
-Sensor::ptr SensorSuite::get(const std::string& uuid) const {
-  return (sensors_.at(uuid));
+void SensorSuite::remove(sensor::Sensor& sensor) {
```
`const sensor::Sensor&`.

Check the other places, see if there are more cases that need `const`.

vauduong commented on src/esp/sensor/Sensor.cpp:
```diff
@@ -51,9 +50,30 @@ Sensor::Sensor(scene::SceneNode& node, SensorSpec::ptr spec)
   setTransformationFromSpec();
 }
 
+Sensor::~Sensor() {
```
Some problems: When a SceneNode which contains a feature is deleted, it seems that the feature is released so that calling node() segfaults, and we can't remove the reference from relevant SensorSuites.

vauduong replied:
@mosra - could you take a look at this destructor? Specifically, we are having some problems with the order of destructors. We want the destructor for the Sensor to be called before its node is released, but it seems to be the other way around.

mosra replied:
> We want the destructor for the Sensor to be called before its node is released, but it seems to be the other way around.

It's ... both :) The destruction order for the `SceneNode` is the following:

1. `~SceneNode()` gets called
2. all `SceneNode` member destructors get called
3. `~Object()` gets called
4. `~Object()` iterates through all attached features and `delete`s these
5. `~Object()` iterates through all children and `delete`s these (or maybe this is first and then the features, not sure)

I don't know the backtrace of the crash, but according to what I see done here, I assume it's because at step (4) you're calling `getSubtreeSensorSuite()`, which however gives you a reference to a member that got already destructed at step (2).

Magnum itself can't really do anything about this since at the point the `~Object()` gets called, all subclass state is already gone. 

*[~~30~~ 60 minutes of head scratching pass]*

I see ~~two~~ three options:

- You'd explicitly delete all attached Sensor features in `~SceneNode()`, which basically puts the step (4) before step (2) and it shouldn't crash anymore ~~i think~~ i hope. Not 100% sure in what state the parent nodes would be at that point tho, maybe those are already deleted as well so this only moves the problem one level up ... would need *extensive* tests. Or maybe you go through all children there as well and delete their sensors too? Sounds heavy.
- Or you do a `dynamic_cast<scene::SceneNode*>(current)` and then check the returned pointer. If it's null, then the parent node got already destructed and thus there's no point in cleaning up anything, as all the state is gone already anyway; if it's non-null then the parent is still alive and so it makes sense to do the `remove()`. Though here you need to explicitly test that when you delete just a subtree, the rest of the graph won't be left with outdated information. I think that should be avoided if you continue going up the parent hierarchy even when `dynamic_cast` returns `nullptr` (instead of for example `break`ing after that), but again, that needs an explicit test.
- Or rethink how all this information is managed and what's responsible for updating the tree when operations like deleting a subtree happen. Doesn't seem like a task for one afternoon, though.

@bigbike, this is exactly what i feared of when we were discussing this a few weeks ago, that keeping all this redundant information up-to-date will be a huge and complex task prone to subtle errors :sweat_smile:

bigbike replied:
@mosra: by the way, this is the backtrace:

```
#0  0x000055555572d18d in Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<float> >::parent (this=0x0)
    at /home/yili/temp/habitat-sim/src/deps/magnum/src/Magnum/SceneGraph/Object.h:151
#1  esp::sensor::Sensor::~Sensor (this=0x5555561d2fe0, __in_chrg=<optimised out>)
    at /home/yili/temp/habitat-sim/src/esp/sensor/Sensor.cpp:70
#2  0x000055555572c43c in esp::sensor::VisualSensor::~VisualSensor (this=0x5555561d2fe0, 
    __in_chrg=<optimised out>) at /usr/include/c++/9/bits/move.h:99
#3  esp::sensor::CameraSensor::~CameraSensor (this=0x5555561d2fe0, __in_chrg=<optimised out>)
    at /home/yili/temp/habitat-sim/src/esp/sensor/CameraSensor.h:25
#4  esp::sensor::CameraSensor::~CameraSensor (this=0x5555561d2fe0, __in_chrg=<optimised out>)
    at /home/yili/temp/habitat-sim/src/esp/sensor/CameraSensor.h:25
#5  0x000055555572c088 in Corrade::Containers::LinkedListItem<Magnum::SceneGraph::AbstractFeature<3u, float>, Magnum::SceneGraph::AbstractObject<3u, float> >::doErase (this=<optimised out>)
    at /home/yili/temp/habitat-sim/src/deps/corrade/src/Corrade/Containers/LinkedList.h:345
#6  0x0000555555850633 in Corrade::Containers::LinkedListItem<Magnum::SceneGraph::AbstractFeature<3u, float>, Magnum::SceneGraph::AbstractObject<3u, float> >::erase (this=<optimised out>)
    at /home/yili/temp/habitat-sim/src/deps/corrade/src/Corrade/Containers/LinkedList.h:419
#7  Corrade::Containers::LinkedList<Magnum::SceneGraph::AbstractFeature<3u, float> >::erase (
    item=<optimised out>, this=0x5555561d2da8)
    at /home/yili/temp/habitat-sim/src/deps/corrade/src/Corrade/Containers/LinkedList.h:381
#8  Corrade::Containers::LinkedList<Magnum::SceneGraph::AbstractFeature<3u, float> >::clear (
    this=0x5555561d2da8)
    at /home/yili/temp/habitat-sim/src/deps/corrade/src/Corrade/Containers/LinkedList.h:389
#9  Corrade::Containers::LinkedList<Magnum::SceneGraph::AbstractFeature<3u, float> >::~LinkedList (
    this=0x5555561d2da8, __in_chrg=<optimised out>)
    at /home/yili/temp/habitat-sim/src/deps/corrade/src/Corrade/Containers/LinkedList.h:123
#10 Magnum::SceneGraph::AbstractObject<3u, float>::~AbstractObject (this=this@entry=0x5555561d2da0, 
    __in_chrg=<optimised out>)
    at /home/yili/temp/habitat-sim/src/deps/magnum/src/Magnum/SceneGraph/Object.hpp:42
#11 0x0000555555868b68 in Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<float> >::~Object (this=0x5555561d2da0, __in_chrg=<optimised out>)
    at /home/yili/temp/habitat-sim/src/deps/corrade/src/Corrade/Containers/EnumSet.h:98
#12 0x000055555573c33b in non-virtual thunk to esp::scene::SceneNode::~SceneNode() ()
    at /home/yili/temp/habitat-sim/src/deps/magnum/src/Magnum/SceneGraph/Object.h:358
#13 0x0000555555728606 in Corrade::Containers::LinkedListItem<Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<float> >, Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<float> > >::doErase (
    this=<optimised out>)
    at /home/yili/temp/habitat-sim/src/deps/corrade/src/Corrade/Containers/LinkedList.h:345
#14 0x0000555555868b10 in Corrade::Containers::LinkedListItem<Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<float> >, Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<float> > >::erase (this=<optimised out>)
    at /home/yili/temp/habitat-sim/src/deps/corrade/src/Corrade/Containers/LinkedList.h:419
#15 Corrade::Containers::LinkedList<Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<float> > >::erase (item=<optimised out>, this=0x5555561d0e68)
    at /home/yili/temp/habitat-sim/src/deps/corrade/src/Corrade/Containers/LinkedList.h:381
#16 Corrade::Containers::LinkedList<Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<float> > >::clear (this=0x5555561d0e68)
    at /home/yili/temp/habitat-sim/src/deps/corrade/src/Corrade/Containers/LinkedList.h:389
#17 Corrade::Containers::LinkedList<Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<float> > >::~LinkedList (this=0x5555561d0e68, 
    __in_chrg=<optimised out>)
    at /home/yili/temp/habitat-sim/src/deps/corrade/src/Corrade/Containers/LinkedList.h:123
#18 Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<fl--Type <RET> for more, q to quit, c to continue without paging--
oat> >::~Object (this=0x5555561d0e20, __in_chrg=<optimised out>)
    at /home/yili/temp/habitat-sim/src/deps/magnum/src/Magnum/SceneGraph/Object.h:105
#19 0x000055555573c33b in non-virtual thunk to esp::scene::SceneNode::~SceneNode() ()
    at /home/yili/temp/habitat-sim/src/deps/magnum/src/Magnum/SceneGraph/Object.h:358
#20 0x0000555555728606 in Corrade::Containers::LinkedListItem<Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<float> >, Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<float> > >::doErase (
    this=<optimised out>)
    at /home/yili/temp/habitat-sim/src/deps/corrade/src/Corrade/Containers/LinkedList.h:345
#21 0x0000555555868b10 in Corrade::Containers::LinkedListItem<Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<float> >, Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<float> > >::erase (this=<optimised out>)
    at /home/yili/temp/habitat-sim/src/deps/corrade/src/Corrade/Containers/LinkedList.h:419
#22 Corrade::Containers::LinkedList<Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<float> > >::erase (item=<optimised out>, this=0x5555561d0bf8)
    at /home/yili/temp/habitat-sim/src/deps/corrade/src/Corrade/Containers/LinkedList.h:381
#23 Corrade::Containers::LinkedList<Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<float> > >::clear (this=0x5555561d0bf8)
    at /home/yili/temp/habitat-sim/src/deps/corrade/src/Corrade/Containers/LinkedList.h:389
#24 Corrade::Containers::LinkedList<Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<float> > >::~LinkedList (this=0x5555561d0bf8, 
    __in_chrg=<optimised out>)
    at /home/yili/temp/habitat-sim/src/deps/corrade/src/Corrade/Containers/LinkedList.h:123
#25 Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<float> >::~Object (this=0x5555561d0bb0, __in_chrg=<optimised out>)
    at /home/yili/temp/habitat-sim/src/deps/magnum/src/Magnum/SceneGraph/Object.h:105
#26 0x000055555573c33b in non-virtual thunk to esp::scene::SceneNode::~SceneNode() ()
    at /home/yili/temp/habitat-sim/src/deps/magnum/src/Magnum/SceneGraph/Object.h:358
#27 0x0000555555728606 in Corrade::Containers::LinkedListItem<Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<float> >, Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<float> > >::doErase (
    this=<optimised out>)
    at /home/yili/temp/habitat-sim/src/deps/corrade/src/Corrade/Containers/LinkedList.h:345
#28 0x0000555555868b10 in Corrade::Containers::LinkedListItem<Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<float> >, Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<float> > >::erase (this=<optimised out>)
    at /home/yili/temp/habitat-sim/src/deps/corrade/src/Corrade/Containers/LinkedList.h:419
#29 Corrade::Containers::LinkedList<Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<float> > >::erase (item=<optimised out>, this=0x5555561cfa30)
    at /home/yili/temp/habitat-sim/src/deps/corrade/src/Corrade/Containers/LinkedList.h:381
#30 Corrade::Containers::LinkedList<Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<float> > >::clear (this=0x5555561cfa30)
    at /home/yili/temp/habitat-sim/src/deps/corrade/src/Corrade/Containers/LinkedList.h:389
#31 Corrade::Containers::LinkedList<Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<float> > >::~LinkedList (this=0x5555561cfa30, 
    __in_chrg=<optimised out>)
    at /home/yili/temp/habitat-sim/src/deps/corrade/src/Corrade/Containers/LinkedList.h:123
#32 Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<float> >::~Object (this=0x5555561cf9e8, __in_chrg=<optimised out>)
    at /home/yili/temp/habitat-sim/src/deps/magnum/src/Magnum/SceneGraph/Object.h:105
#33 0x00005555557278d7 in esp::scene::SceneNode::~SceneNode (this=0x5555561cf9e8, 
    __in_chrg=<optimised out>)
    at /home/yili/temp/habitat-sim/src/deps/corrade/src/Corrade/Containers/Optional.h:240
#34 esp::scene::SceneGraph::~SceneGraph (this=0x5555561cf960, __in_chrg=<optimised out>)
--Type <RET> for more, q to quit, c to continue without paging--
    at /home/yili/temp/habitat-sim/src/esp/scene/SceneGraph.h:25
#35 esp::scene::SceneGraph::~SceneGraph (this=0x5555561cf960, __in_chrg=<optimised out>)
    at /home/yili/temp/habitat-sim/src/esp/scene/SceneGraph.h:25
#36 std::default_delete<esp::scene::SceneGraph>::operator() (this=0x5555561d0b90, 
    __ptr=0x5555561cf960) at /usr/include/c++/9/bits/unique_ptr.h:81
#37 std::unique_ptr<esp::scene::SceneGraph, std::default_delete<esp::scene::SceneGraph> >::~unique_ptr (this=0x5555561d0b90, __in_chrg=<optimised out>) at /usr/include/c++/9/bits/unique_ptr.h:292
#38 std::_Destroy<std::unique_ptr<esp::scene::SceneGraph, std::default_delete<esp::scene::SceneGraph> > > (__pointer=0x5555561d0b90) at /usr/include/c++/9/bits/stl_construct.h:98
#39 std::_Destroy_aux<false>::__destroy<std::unique_ptr<esp::scene::SceneGraph, std::default_delete<esp::scene::SceneGraph> >*> (__last=<optimised out>, __first=0x5555561d0b90)
    at /usr/include/c++/9/bits/stl_construct.h:108
#40 std::_Destroy<std::unique_ptr<esp::scene::SceneGraph, std::default_delete<esp::scene::SceneGraph> >*> (__last=<optimised out>, __first=<optimised out>)
    at /usr/include/c++/9/bits/stl_construct.h:137
#41 std::_Destroy<std::unique_ptr<esp::scene::SceneGraph, std::default_delete<esp::scene::SceneGraph> >*, std::unique_ptr<esp::scene::SceneGraph, std::default_delete<esp::scene::SceneGraph> > > (
    __last=0x5555561d0b98, __first=<optimised out>) at /usr/include/c++/9/bits/stl_construct.h:206

#42 std::vector<std::unique_ptr<esp::scene::SceneGraph, std::default_delete<esp::scene::SceneGraph> >, std::allocator<std::unique_ptr<esp::scene::SceneGraph, std::default_delete<esp::scene::SceneGraph> > > >::~vector (this=0x7fffffffb300, __in_chrg=<optimised out>)
    at /usr/include/c++/9/bits/stl_vector.h:677
#43 esp::scene::SceneManager::~SceneManager (this=0x7fffffffb300, __in_chrg=<optimised out>)
    at /home/yili/temp/habitat-sim/src/esp/scene/SceneManager.h:25
#44 SensorTest::testSensorFactory (this=<optimised out>)
    at /home/yili/temp/habitat-sim/src/tests/SensorTest.cpp:31
#45 0x000055555574cc7d in Corrade::TestSuite::Tester::exec (this=0x7fffffffd9f0, 
    previousTester=<optimised out>, logOutput=<optimised out>, errorOutput=<optimised out>)
    at /home/yili/temp/habitat-sim/src/deps/corrade/src/Corrade/TestSuite/Tester.cpp:476
#46 0x000055555570cda6 in main (argc=<optimised out>, argv=<optimised out>)
    at /home/yili/temp/habitat-sim/src/tests/SensorTest.cpp:115

```

mosra replied:
Ah. That's because you basically do 

```
node().parent()->parent()
               ^
               *boom*, this might be null aready
```

vauduong replied:
@bigbike @mosra Thank you so much for such an in-depth response!
Yes, I'm finding that node() is actually already null when the destructor is called. This info is really good am I'm trying to see if I can find a workaround.

mosra replied:
Just for the record after a private discussion with Yili -- the problem was a subtlety in the behavior of LinkedList, which is what the objects and features use to connect to each other -- and should be fixed with https://github.com/mosra/corrade/commit/1daaac6ef5b8aef9eaa31f1def8a8f26c5ca8906, which has https://github.com/facebookresearch/habitat-sim/pull/1123 as a prerequisite.

bigbike replied:
> Thanks for the fix to Magnum! That's works

@vauduong : When you said this, are you sure you applied:
[mosra/corrade@1daaac6](https://github.com/mosra/corrade/commit/1daaac6ef5b8aef9eaa31f1def8a8f26c5ca8906)
Please goto the corrade's source code to check. It is in the src/deps/corrade

>current from auto current = node().parent() will just be nullptr.

@mosra : This worries me a lot. You might be worried too. because the "half-destroyed" theory (SceneNode is destroyed, but BASE is not destroyed.) may not be true (assuming Vivian already applied 1daaac6.

vauduong commented on src/tests/SensorTest.cpp:
```diff
@@ -29,74 +29,85 @@ SensorTest::SensorTest() {
+  // Remove sensor from parent node and assert that it is no longer in parent
+  // subtreeSensorSuite
+  parentNode.getSubtreeSensorSuite().get("A").~Sensor();
+  CORRADE_VERIFY(parentNode.getSubtreeSensorSuite().getSensors().size() ==
+                 7);  // 6
+
+  // Remove sensor from child node and assert that it is no longer in parent and
+  // child subtreeSensorSuite
+  childNode.getSubtreeSensorSuite().get("D").~Sensor();
+  CORRADE_VERIFY(parentNode.getSubtreeSensorSuite().getSensors().size() ==
+                 7);  // 5
+  CORRADE_VERIFY(childNode.getSubtreeSensorSuite().getSensors().size() ==
+                 3);  // 2
+
+  // Delete child node and assert that sensors are no longer in parentNode
+  childNode.~SceneNode();
+  CORRADE_VERIFY(parentNode.getSubtreeSensorSuite().getSensors().size() ==
+                 7);  // 3
 }
```
These tests currently have incorrect values and will be fixed.

vauduong commented on src/tests/CMakeLists.txt:
```diff
@@ -15,22 +15,8 @@ configure_file(
   ${CMAKE_CURRENT_SOURCE_DIR}/configure.h.cmake ${CMAKE_CURRENT_BINARY_DIR}/configure.h
 )
 
-test(AttributesManagersTest assets metadata)
-target_include_directories(AttributesManagersTest PRIVATE ${CMAKE_CURRENT_BINARY_DIR})
-
```
These tests have been temporarily removed for faster debugging. They all pass and will be included in the final PR! :)

bigbike reviewed (commented):
Oh no. You must committed the 3rd-party library by accident.

Please revert these changes.

bigbike commented on src/esp/sensor/SensorFactory.h:
```diff
@@ -6,16 +10,30 @@ namespace sensor {
+   * specs for each sensor
+   *
+   */
+  static void createSensors(scene::SceneNode& node,
```
we can return a `std::vector<std::reference_wrapper<Sensor>>`

bigbike commented on src/tests/SensorTest.cpp:
```diff
@@ -19,84 +19,159 @@ struct SensorTest : Cr::TestSuite::Tester {
+  CORRADE_VERIFY(grandchildNode.getNodeSensorSuite().getSensors().size() == 3);
+  CORRADE_VERIFY(grandchildNode.getSubtreeSensorSuite().getSensors().size() ==
+                 3);
+
```
You can try to delete a sensor that never exists.

vauduong replied:
Sounds good, I can catch the exception.

bigbike commented on src/tests/SensorTest.cpp:
```diff
@@ -19,84 +19,159 @@ struct SensorTest : Cr::TestSuite::Tester {
+  CORRADE_VERIFY(grandchildNode.getSubtreeSensorSuite().getSensors().size() ==
+                 3);
+
+  // Remove child node and assert grandchild node is destructed as well
```
you can create a more complex scene graph, and delete a subtree.

bigbike commented on src/esp/scene/SceneNode.h:
```diff
@@ -41,6 +58,7 @@ class SceneNode : public MagnumObject,
   // terminate node (e.g., "MagnumScene" defined in SceneGraph) as its ancestor
   SceneNode() = delete;
   SceneNode(SceneNode& parent);
+  ~SceneNode() override;
```
By the way, I know in the base class, there is a function called `setParent(parentNode)` which moves a subtree rooted at "this" node to another node `parentNode`.

Certainly you need to overload that function as well. In this function:
- You need to check if the parentNode is a leaf node.
- Update the sensorSuites in the old ancestors
- setParent
- Update the sensorSuites in the new ancestors

This is quite challenging. Please think very carefully:
- when "this" node is a `leaf` node.
- when "this" node is NOT a `leaf` node.

bigbike replied:
Be CAREFUL! No duplicate code.
Expect some code will be used in both ctor and setParent, also dctor and setParent. 
You need to extract a function to accommodate such code.

bigbike commented on src/esp/scene/SceneNode.h:
```diff
@@ -88,6 +118,15 @@ class SceneNode : public MagnumObject,
   const Magnum::Range3D& getCumulativeBB() const { return cumulativeBB_; };
 
+  // Return SensorSuite containing references to Sensors this SceneNode holds
+  esp::sensor::SensorSuite& getNodeSensorSuite() { return *nodeSensorSuite_; }
```
We should also provide another 2 wrapper functions for convenience:
`getNodeSensors()` which will return `nodeSensorSuite_->getSensors()`
and 
`getSubtreeSensor()` ...

erikwijmans commented on src/esp/scene/SceneNode.cpp:
```diff
@@ -16,16 +18,61 @@ SceneNode::SceneNode(SceneNode& parent)
+  // First erase all nodes in nodeSensorSuite from its parent's nodeSensorSuite
+  for (const auto& sensor : nodeSensorSuite_->getSensors()) {
+    static_cast<SceneNode*>(this->parent())
+        ->getNodeSensorSuite()
```
Should this be the subtreeSensorSuite?

bigbike replied:
@erikwijmans : in the `nodeSensorSuite`, it stores the sensors owned by this node. For example, if this node represents "left arm" and owns 3 sensors (in 3 child, LEAF nodes), then these 3 sensors will be in this `nodeSensorSuite` and it does not include sensors in the subtrees, e.g., sensors on the "fingers". 

Now suppose you are to delete 1 of the 3 sensors. We do so by deleting that sensor node (the sensor "holder" node). In this case, this for loop is to help get rid of the sensor in its actual "owner".

erikwijmans replied:
I think I get how this works without being a nightmare after looking more in depth.  You are stopping a sensor from being attached to a sensor by tagging that scene node as a permanent leaf.  

I am trying to think of how to make this more clear in the code.  Maybe adding some checks that must be true due to this logic?  If I am understanding this correctly, this loop is unnecessary for non-leaves as a node's parent will only have its child sensor in its sensor suite if that child is a leaf?  A corollary is that non-leaves do not own sensors do not own sensors, so the sensorSuite for a non-leaf should be empty upon destruction?

vauduong replied:
Yes, that's correct. 
- Sensors can only be attached to leaf nodes.
- For non-leaves, the loop is unnecessary because it will own no sensors that could be in its parent's nodeSensorSuite.
- Non-leaves do not own sensors, but sensors could be attached to its direct child nodes, in which case they will be included in its nodeSensorSuite. (Otherwise, nodeSensorSuites would only ever contain 0 or 1 sensors)

bigbike commented on src/esp/scene/SceneNode.cpp:
```diff
@@ -16,16 +18,61 @@ SceneNode::SceneNode(SceneNode& parent)
+  // In case 2, you need update the sensorSuites stored in each ancestor node in
+  // a bottom-up manner.
+
+  // First erase all nodes in nodeSensorSuite from its parent's nodeSensorSuite
```
@vauduong :  I thought about it. The following for loop is actually only needed for the LEAF nodes. Is it correct?

If yes, we can add a if clause.
If no, let me know your thoughts.

vauduong replied:
Yes, I think this is true. 
If it's a leaf node, that means its parent node will have its sensor in the nodeSensorSuite.
If it's not a leaf node, then that means any sensors in its nodeSensorSuite are child Sensors, and will just be deleted when the node is deleted.

bigbike replied:
If node "A" is not a leaf node, any sensor in the `nodeSensorSuite_` of "A" will not be in the nodeSensorSuite_ of "A"'s parent for sure.

bigbike commented on src/esp/sensor/Sensor.cpp:
```diff
@@ -47,13 +47,29 @@ void SensorSpec::sanityCheck() {
 Sensor::Sensor(scene::SceneNode& node, SensorSpec::ptr spec)
     : Magnum::SceneGraph::AbstractFeature3D{node}, spec_(std::move(spec)) {
   node.setType(scene::SceneNodeType::SENSOR);
+  node.setSceneNodeTags({scene::SceneNodeTag::Leaf});
```
Here, you need to CORRADE_ASSERT the node does NOT have any child before adding the tag.
`"Sensor::Sensor(): Cannot attach a sensor to a non-LEAF node. The number of children of this node is not zero."`

vauduong commented on src/esp/sensor/SensorFactory.cpp:
```diff
@@ -28,7 +31,16 @@ sensor::SensorSuite SensorFactory::createSensors(
+
+void SensorFactory::deleteSensor(scene::SceneNode& node,
+                                 const std::string& uuid) {
+  delete (&node.getSubtreeSensorSuite().get(uuid).node());
```
Currently std::map will throw an out of range exception, so it's up to the user to make sure that sensors are not prematurely deleted.

bigbike replied:
> so it's up to the user to make sure

No, this is not a good idea. As a developer, one of our duties is to keep the system robust so that you should handle the illegal cases.
Here, the correct way is that if such a sensor (with uuid) does not exist, the code will just return directly (I never say using ASSERT), no harm to the system.

bigbike commented on src/esp/scene/SceneNode.cpp:
```diff
@@ -16,16 +18,61 @@ SceneNode::SceneNode(SceneNode& parent)
+  subtreeSensorSuite_ = new esp::sensor::SensorSuite(*this);
+}
+
+SceneNode::~SceneNode() {
```
I just had a meeting with @mosra this morning.
He found the root cause causing `node().parent()` is nullptr.
After he push the fixes, he will bump the habitat to a newer version of corrade and magnum.
Then we will move the code back to the `~Sensor()`.
Yeah, it is a bit inconvenience, and annoying. But the reason is that: We inserted a lot of logic in the Sensor's ctor. Whatever constructed in the Sensor's ctor, should be deconstructed in the Sensor's dtor. In terms of design, it is MUCH LESS error-prone.

bigbike commented on src/esp/scene/SceneNode.cpp:
```diff
@@ -16,16 +18,61 @@ SceneNode::SceneNode(SceneNode& parent)
+  }
 }
 
 SceneNode& SceneNode::createChild() {
```
Also upgrade this function to:
`createChild(SceneNodeTag childNodeTag = {})`
So when you create sensors in the SensorFactory, you can directly set the tag even earlier during the construction.

vauduong commented:
Still have more work to do on the python side, python compiler is having some trouble finding magnum on local machine so have not tested python yet

vauduong commented:
@bigbike @erikwijmans @Skylion007 
Hello! Could you all provide some feedback on changes to the python when you get the chance?

In particular, agent.py now uses SensorFactory to initialize sensors. Wrapper classes Sensor and SensorSuite are now deprecated on the python side, and sensors are kept track of on the C++ side using the node and subtreeSensorSuites.

Let me know if you have any questions about this!

vauduong commented on habitat_sim/simulator.py:
```diff
@@ -341,21 +318,153 @@ def get_sensor_observations(
+            render_flags |= habitat_sim.gfx.Camera.Flags.OBJECTS_ONLY
+            self.renderer.draw(sensor, self.get_active_scene_graph(), render_flags)
+
+    def get_observation(self, sensor: Sensor) -> Union[ndarray, "Tensor"]:
```
Thanks Erik! This was definitely a concern when I was refactoring, but I wasn't sure how to address it with the C++ handling sensor storage now. The hook class makes a lot of sense and I will work on implementing it!

vauduong commented on src/esp/scene/SceneNode.cpp:
```diff
@@ -12,26 +14,148 @@ namespace scene {
+
+  // First remove all sensors in nodeSensorSuite from its parent's
+  // nodeSensorSuite if node is leaf node
+  if (getSceneNodeTags() &= SceneNodeTag::Leaf) {
```
Thanks for the catch!

erikwijmans commented on habitat_sim/simulator.py:
```diff
@@ -473,28 +445,23 @@ def step_physics(self, dt: float, scene_id: int = 0) -> None:
-    def __init__(self, sim: Simulator, agent: Agent, sensor_id: str) -> None:
-        self._sim = sim
-        self._agent = agent
+class SensorReader:
```
IMO it still makes sense to have the simulator drive the drawing process.  This class can then be mostly stateless and dumb then.

vauduong replied:
Thanks Erik, that's fixed in the newest PR. I'm having some trouble with "SensorReader" being undefined though, and I'm not sure how to fix it. Do you have any suggestions on what I'm missing?

erikwijmans replied:
I think you need to put the SensorReader name in quotes for the typing as that symbol hasn't been defined yet

vauduong commented:
TODO: Keep fixing python implementation, add baseline tests to check that SensorSuites are correct for setParent() and Sensor deletion, per @mathfac 's suggestions

erikwijmans commented on habitat_sim/simulator.py:
```diff
@@ -472,24 +448,115 @@ def __del__(self) -> None:
 
-        self._sim.renderer.bind_render_target(self._sensor_object)
+        self._sim = sim
+        self._sensor_object: Sensor
```
Ownership is gonna get messy here since the scene graph owns the sensor.  If the sensor reader doesn't keep a reference to the sensor, that'll make it much cleaner

vauduong replied:
Ok, that can be done!

erikwijmans commented on habitat_sim/simulator.py:
```diff
@@ -472,24 +448,115 @@ def __del__(self) -> None:
     def step_physics(self, dt: float, scene_id: int = 0) -> None:
         self.step_world(dt)
 
+    def draw_observation(self, sensor_reader: SensorReader) -> None:
```
Using the SensorReader here and in `get_observation` seems wrong.  The reader is an implementation detail to cache things, not the driver.  The sensor is what drives.

bigbike replied:
I totally agree with you. @erikwijmans

bigbike replied:
> The sensor is what drives.
@erikwijmans : when you said this, my understanding was it should be: `sensor.draw_observation()`. Correct?

vauduong replied:
Thanks for the insight Erik, the torch buffer type is a big concern. Based on what you've said here, it seems we will have to keep the Sensor wrapper class in python, and subsequently, keep the global Dict[str, Sensor] in simulator as well. Although it might make sense to think of the python Sensor as a SensorReader, it seems this might not be something the users would need to know.

Skylion007 replied:
@erikwijmans @vauduong . I think I figured out hwo to read into a Torch / Numpy buffer directly from C++. PyBind11 has a C++ Wrapper for Python meaning you can execute Python code in the generated functions (including importing pure python modules at runtime).  https://pybind11.readthedocs.io/en/stable/advanced/pycpp/object.html#accessing-python-libraries-from-c . We could do attach the minimal pure python functions to the Sensor Class using that.

bigbike commented on habitat_sim/simulator.py:
```diff
@@ -341,16 +322,17 @@ def get_sensor_observations(
-            for _sensor_uuid, sensor in agent_sensorsuite.items():
-                sensor.draw_observation()
+            for uuid in self.get_agent(agent_id).scene_node.node_sensors.items():
+                self.draw_observation(self._sensor_readers[uuid])
```
May I understand the purpose to build a simulator level "draw_observation"? Such usage seems unnatural to me.
I think it still makes sense to call `sensor.draw_observation()`. The only difference is that where the sensors come from. Before, a sensor is from the sensorSuite in the Agent class. Now it should be from the node where the Agent resides in.

vauduong replied:
Yes, I think it makes sense that the sensor should draw the observation as well. Sorry, from earlier feedback that the simulator should drive the drawing process, I was under the impression that simulator should own the function to draw the observation, while SensorReader would just own the implementation to initialize the noise model and buffers. https://github.com/facebookresearch/habitat-sim/pull/1118#discussion_r590851442

bigbike commented on habitat_sim/agent/agent.py:
```diff
@@ -133,19 +132,16 @@ class Agent(object):
         self,
         scene_node: hsim.SceneNode,
         agent_config: Optional[AgentConfiguration] = None,
-        _sensors: Optional[SensorSuite] = None,
```
Great. It is not needed as an input parameter any more.

bigbike commented on habitat_sim/agent/agent.py:
```diff
@@ -133,19 +132,16 @@ class Agent(object):
         controls: Optional[ObjectControls] = None,
     ) -> None:
         self.agent_config = agent_config if agent_config else AgentConfiguration()
-        self._sensors = _sensors if _sensors else SensorSuite()
```
- We can keep the concept `_sensors` in the agent class, a convenient alias to the sensorSuite in the node. 
- For an atomic agent, which is the current case, ONLY one `_sensors` is needed and it serves as a reference to the `subtreeSensorSuite` of that node. You should do proper assignment here: `self._sensors = call getter function to obtain the subtreeSensorSuite of that node`
- For an articulated agent, which will be the case in the future, there would be two references, one is `_node_owned_sensors` and `_subtree_sensors` (for example, at the root node of this articulated agent, the `_subtree_sensors` means all the sensors being attached to this agent and the `_node_owned_sensors` are sensors on the "head" assuming the head node is the root node of this articulated agent.)

bigbike commented on habitat_sim/agent/agent.py:
```diff
@@ -166,31 +162,17 @@ def reconfigure(
         self.agent_config = agent_config
 
         if reconfigure_sensors:
-            self._sensors.clear()
```
See my previous comments.
If you keep the `_sensors` concept, you do not have to change the code here.

bigbike commented on habitat_sim/sensors/sensor_suite.py:
```diff
@@ -1,12 +0,0 @@
-from habitat_sim import bindings as hsim
-
-
-class SensorSuite(Dict[str, hsim.Sensor]):
```
Great. It should be gone.

bigbike commented on habitat_sim/simulator.py:
```diff
@@ -70,9 +70,9 @@ class Simulator(SimulatorBackend):
+    _sensor_readers: Dict[str, SensorReader] = attr.ib(factory=dict, init=False)
     _num_total_frames: int = attr.ib(default=0, init=False)
     _default_agent_id: int = attr.ib(default=0, init=False)
-    __sensors: List[Dict[str, "Sensor"]] = attr.ib(factory=list, init=False)
```
Yes, it should be gone.

bigbike commented on habitat_sim/simulator.py:
```diff
@@ -118,21 +118,6 @@ def __attrs_post_init__(self) -> None:
-            agent.close()
-            del agent
-
-        self.agents = []
```
It deletes all the agents.
Are we sure we can remove this piece of code?

bigbike commented on habitat_sim/simulator.py:
```diff
@@ -254,19 +239,10 @@ def __set_from_config(self, config: Configuration) -> None:
 
         self._default_agent_id = config.sim_cfg.default_agent_id
 
-        self.__sensors: List[Dict[str, Sensor]] = [
```
Yes, it should be gone.

bigbike commented on habitat_sim/simulator.py:
```diff
@@ -254,19 +239,10 @@ def __set_from_config(self, config: Configuration) -> None:
-        self.__sensors[agent_id][uuid] = Sensor(
-            sim=self, agent=self.get_agent(agent_id), sensor_id=uuid
-        )
+                self.add_sensor(spec, agent_id=agent_id)
```
The argument order is really weird. 
The natural way should be agent id first, then the spec.

bigbike commented on habitat_sim/simulator.py:
```diff
@@ -294,8 +270,13 @@ def add_sensor(
-        agent._add_sensor(sensor_spec)
-        self._update_simulator_sensors(sensor_spec.uuid, agent_id=agent_id)
+        agent._add_sensor(sensor_spec, modify_agent_config=False)
+        self._sensor_readers[sensor_spec.uuid] = SensorReader(
```
Wait, what is the `_sensor_readers[sensor_spec.uuid]`? 
May I understand what is the purpose to have such a `SensorReader` class?
At a glance, this `_sensor_readers` serves as a global dict again to me, which is a RED flag.
The main purpose of this PR is to remove such kind of global lookup table, hiding them in the C++ library (e.g., we move and manage the SensorSuite into the node class), what you had here is doing it in the opposite way. 

Again, I might be wrong. But I would like you to justify such design.

vauduong replied:
To explain what I'm doing here:
I removed the global dict _sensors in simulator as now sensors are being kept track of by the SensorSuites. With this I also removed any code to update _sensors, including `update_simulator_sensors`, which initializes the Sensor wrapper.

Python has its own implementation to get observations, which requires initializing a Sensor wrapper which contains torch/numpy buffer and noise model for sensors and storing it in agent_observations. 

In order to make the Sensor wrapper more intuitive to understand, I renamed it to SensorReader, because all it really does is initialize the buffer and noise model and provide its own implementation to get observations. Then we needed a _sensor_readers global dict to hold these buffers and noise models, so functionally nothing has really changed.

However, I realize this is not the approach we want to go for. We really don't want this global dict.

bigbike commented on habitat_sim/simulator.py:
```diff
@@ -341,16 +322,17 @@ def get_sensor_observations(
-            for sensor_uuid, sensor in self.__sensors[agent_id].items():
-                agent_observations[sensor_uuid] = sensor.get_observation()
+            for uuid in self.get_agent(agent_id).scene_node.node_sensors.items():
+                agent_observations[uuid] = self.get_observation(
```
again, why not `sensor.get_observation()` ?

bigbike commented on habitat_sim/simulator.py:
```diff
@@ -472,24 +448,115 @@ def __del__(self) -> None:
+                 (has it been detached from a scene node?)"
+            )
 
-class Sensor:
```
May I understand why the `Sensor` class is gone?

bigbike commented:
@vauduong :
I am a bit confused. Based on our plan, the changes on the py side is another milestone project and is supposed to be done in another PR. May I understand the reason why you put them in this PR?

Before you introduced the py changes, everything in this PR resided on the C++ side. And it is brand new functionality in the C++ simulator, so I do not think it is a breaking changes in terms of the python users. Yes, you removed the sensorSuite from the C++ `Agent` class, but it should not affect the py `Agent` class since it has its own, independent `sensorSuite` implementation.

It is possible I have missed something so that my opinions above could be wrong. Let me know.

Also looking forward to seeing your answers to all my questions in the diff as they are highly related to the design. Thanks.

vauduong commented:
@bigbike 
Thanks for the feedback! Sorry, I was under the impression that the python implementation should be addressed in this PR as well since that is what I had done previously, but I see how it can make the PR harder to review. 

I'll address your questions, put the python changes in another branch, and publish this PR!

vauduong commented on habitat_sim/simulator.py:
```diff
@@ -254,19 +239,10 @@ def __set_from_config(self, config: Configuration) -> None:
 
         self._default_agent_id = config.sim_cfg.default_agent_id
 
-        self.__sensors: List[Dict[str, Sensor]] = [
-            dict() for i in range(len(config.agents))
-        ]
         self.__last_state = dict()
         for agent_id, agent_cfg in enumerate(config.agents):
             for spec in agent_cfg.sensor_specifications:
-                self._update_simulator_sensors(spec.uuid, agent_id=agent_id)
-            self.initialize_agent(agent_id)
-
-    def _update_simulator_sensors(self, uuid: str, agent_id: int) -> None:
-        self.__sensors[agent_id][uuid] = Sensor(
-            sim=self, agent=self.get_agent(agent_id), sensor_id=uuid
-        )
```
simulator_sensors holds wrapper Sensors

vauduong commented:
> To explain what I'm doing here:

> I removed the global dict _sensors in simulator as now sensors are being kept track of by the SensorSuites. With this I also removed any code to update _sensors, including update_simulator_sensors, which initializes the Sensor wrapper.
> 
> Python has its own implementation to get observations, which requires initializing a Sensor wrapper which contains torch/numpy buffer and noise model for sensors and storing it in agent_observations.
> 
> In order to make the Sensor wrapper more intuitive to understand, I renamed it to SensorReader, because all it really does is initialize the buffer and noise model and provide its own implementation to get observations. Then we needed a _sensor_readers global dict to hold these buffers and noise models, so functionally nothing has really changed.
> 
> However, I realize this is not the approach we want to go for. We really don't want this global dict.


Here is an explanation for above. 
So it seems there are two solutions: Trampoline class and global dict/SensorReader. 
We really don't want the global dict, so I think the solution is to use a trampoline class as @erikwijmans suggested, but I'm not sure if it will work. We use SensorFactory to initialize sensors, so we will need to force every Sensor initialized to a trampoline class (by creating helper class `PySensor` in C++ and using py::init_alias), then override its functionality in python. However, I'm not sure where to put the implementations for the overridden functions of PySensor? 

I am also not sure how pybind knows which class to initialize?  For example. if I were to add PySensor to the SensorBindings like so, would it be initialized with getObservation and drawObservation as virtual functions? It seems like I will just have to try this out.

```
py::class_<CameraSensor, Magnum::SceneGraph::PyFeature<CameraSensor>,
             VisualSensor, Magnum::SceneGraph::PyFeatureHolder<CameraSensor>, PySensor>(
      m, "CameraSensor")
      .def(py::init_alias<std::reference_wrapper<scene::SceneNode>,
                          const CameraSensorSpec::ptr&>()) 
```

https://pybind11.readthedocs.io/en/stable/advanced/classes.html#forced-trampoline-class-initialisation

@bigbike @erikwijmans @Skylion007 Could you all give any guidance as to if this understanding is correct? I'm pretty new to the concept of trampoline classes, so I'm not sure if I'm thinking about this correctly.

vauduong commented on src/esp/bindings_js/bindings_js.cpp:
```diff
@@ -204,6 +203,16 @@ EMSCRIPTEN_BINDINGS(habitat_sim_bindings_js) {
       .function("setLocalTransform", &Sensor_setLocalTransform)
       .function("specification", &Sensor::specification);
 
+  em::class_<SensorSuite>("SensorSuite")
+      .smart_ptr<SensorSuite::ptr>("SensorSuite::ptr")
+      .function("add", &SensorSuite::add)
+      .function("merge", &SensorSuite::merge)
+      .function("remove",
+                em::select_overload<void(const Sensor&)>(&SensorSuite::remove))
+      .function("remove", em::select_overload<void(const std::string&)>(
+                              &SensorSuite::remove))
+      .function("clear", &SensorSuite::clear);
+
```
A note about SensorSuite's bindings: Since SensorSuite::get now returns a Sensor& instead of Sensor::ptr, I am unable to write a binding for it as doing so would cause an `Allocating an object of abstract class TypeError`, as Sensor as pure virtual functions.

I did some research on this and found that it's possible to implement pure virtual functions in JavaScript, but that seems like not what we want, as derived classes will have their own implementation. https://emscripten.org/docs/porting/connecting_cpp_and_javascript/embind.html#deriving-from-c-classes-in-javascript

The js code still works without this SensorSuite::get method, as much of the logic for SensorSuites happens within other classes. However, this might be something that we want in the future. Does anyone with more experience in emscripten know if it is possible to bind such a method, and can give some tips to do so?

erikwijmans commented on src/esp/scene/SceneNode.cpp:
```diff
@@ -12,26 +14,146 @@ namespace scene {
+    return *this;
+  }
+
+  // Skip if this is parent of newParent
```
I believe this is being done since this is what Magnum does internally?

vauduong replied:
Yes! This is what magnum is checking internally. Since we update some SensorSuites before calling MagnumObject::setParent(), we have to check again to make sure it's valid to update those SensorSuites.

erikwijmans replied:
Makes sense.  Can you add a comment?

bigbike commented on src/esp/sensor/Sensor.cpp:
```diff
@@ -42,14 +42,36 @@ void SensorSpec::sanityCheck() {
+      .getNodeSensorSuite()
+      .add(*this);
+  node.getSubtreeSensorSuite().add(*this);
+  // Traverse up to root node and add sensor to every subtreeSensorSuite
```
Is that possible to call the functions you wrote? e.g., `addSubtreeSensorsToAncestors`

vauduong replied:
Thank you, good catch!

bigbike commented on src/esp/scene/SceneNode.h:
```diff
@@ -53,10 +72,43 @@ class SceneNode : public MagnumObject,
+   */
+  void removeSceneNodeTag(SceneNodeTag sceneNodeTag) {
+    SceneNodeTags toRemove = {sceneNodeTag};
+    sceneNodeTags_ &= ~toRemove;
```
Cannot be just like this?? `sceneNodeTags_ &= ~sceneNodeTag;`

vauduong replied:
No, unfortunately not, the types are incompatible after the ~ operator is applied.

bigbike replied:
@mosra : Would you take a look? I believe there is some way to do it.

mosra replied:
This usually happens due to `CORRADE_ENUMSET_OPERATORS()` not being defined. But I see the macro later down the file, so in this case it's because the function is defined too early, before the compiler even has a chance to know there are the operators :) Solution is to have just a declaration here and have the actual definition after the macro -- either marked as `inline` directly in the header or in the cpp file.

Though probably the least-effort solution would be writing `sceneNodeTags_ &= ~SceneNodeTags{sceneNodeTag};` and not relying on the existence of the operators macro :))

bigbike replied:
@vauduong : Did you read the comment here?

vauduong replied:
Thanks @mosra !

bigbike commented on src/esp/scene/SceneNode.cpp:
```diff
@@ -12,26 +14,148 @@ namespace scene {
+void SceneNode::addSubtreeSensorsToAncestors() {
+  for (const auto& sensor : subtreeSensorSuite_->getSensors()) {
+    SceneNode* currentNode = dynamic_cast<SceneNode*>(this->parent());
+    while (currentNode != nullptr && !SceneGraph::isRootNode(*currentNode)) {
```
`while (currentNode && ...`

bigbike commented on src/esp/scene/SceneNode.cpp:
```diff
@@ -12,26 +14,148 @@ namespace scene {
+
+  // Skip if this is scene or RootNode (which cannot have parent), or if
+  // newParent is already parent
+  if (isScene() || this->parent() == newParent) {
```
Read the comment you wrote: `Skip if this is scene or RootNode (which cannot have parent)...`
It seems you forget the 2nd case.

erikwijmans reviewed (approved):
I'll let @bigbike give the final go ahead on this, but it overall looks good to me!

bigbike commented on src/esp/scene/SceneNode.cpp:
```diff
@@ -12,26 +14,148 @@ namespace scene {
+  addSubtreeSensorsToAncestors();
+
+  // Add sensors to newParent's nodeSensorSuite
+  if (getSceneNodeTags() & SceneNodeTag::Leaf) {
+    for (const auto& sensor : nodeSensorSuite_->getSensors()) {
+      newParent->getNodeSensorSuite().add(sensor.second);
+    }
+  }
```
Do you repeat the above logic ("Add sensors to newParent's nodeSensorSuite") any other places in the code?
If yes, consider creating a new function. e.g., `addSensorsToParentNodeSensorSuite();`
And inside this function, do verify it if is leaf node first. If not, just return.

vauduong replied:
This function has now been created and called in Sensor's constructor, fyi.

bigbike commented on src/esp/scene/SceneNode.cpp:
```diff
@@ -12,26 +14,148 @@ namespace scene {
+  // If current node is a leaf node, it may contain sensors
+  if (getSceneNodeTags() & SceneNodeTag::Leaf) {
+    // Remove sensors from old parent node's nodeSensorSuite
+    removeSensorsFromParentNodeSensorSuite();
```
You know what, you can put the if clause inside this function.
`if (getSceneNodeTags() & SceneNodeTag::Leaf)`

bigbike commented on src/esp/scene/SceneNode.h:
```diff
@@ -88,6 +140,50 @@ class SceneNode : public MagnumObject,
+  getSubtreeSensors();
+
+  /** @brief Remove sensors in this' nodeSensorSuite from parent's
+   * nodeSensorSuite
```
Add a comment, something like:
Note, you only have to call this function when this node is a leaf node.

bigbike commented on src/esp/scene/SceneNode.cpp:
```diff
@@ -12,26 +14,148 @@ namespace scene {
+    return *this;
+  }
+
+  // Skip if this is parent of newParent
```
// Skip if this is an ancestor of newParent

vauduong commented on src/esp/scene/SceneNode.cpp:
```diff
@@ -12,26 +14,163 @@ namespace scene {
+  return *this;
+}
+
+void SceneNode::addSensorToParentNodeSensorSuite() {
+  // Only add a sensor if this is a leaf node and sensor exists
+  if (!(getSceneNodeTags() & SceneNodeTag::Leaf) ||
+      nodeSensorSuite_->getSensors().empty()) {
+    return;
+  }
+  // There only exists 0 or 1 sensors in a leaf nodeSensorSuite
+  // Get the first one, and if it is valid, add it to parent's nodeSensorSuite
+  std::map<std::string, std::reference_wrapper<sensor::Sensor>>::iterator it =
+      nodeSensorSuite_->getSensors().begin();
+  if (it != nodeSensorSuite_->getSensors().end()) {
+    static_cast<scene::SceneNode*>(this->parent())
+        ->getNodeSensorSuite()
+        .add(it->second);
+  }
+}
+
+void SceneNode::removeSensorFromParentNodeSensorSuite() {
+  // Only add a sensor if this is a leaf node and sensor exists
+  if (!(getSceneNodeTags() & SceneNodeTag::Leaf) ||
+      nodeSensorSuite_->getSensors().empty()) {
+    return;
+  }
+  // There only exists 0 or 1 sensors in a leaf nodeSensorSuite
+  // Get the first one, and if it is valid, remove it from parent's
+  // nodeSensorSuite
+  std::map<std::string, std::reference_wrapper<sensor::Sensor>>::iterator it =
+      nodeSensorSuite_->getSensors().begin();
+  if (it != nodeSensorSuite_->getSensors().end()) {
+    static_cast<scene::SceneNode*>(this->parent())
+        ->getNodeSensorSuite()
+        .remove(it->first);
+  }
+}
+
```
Thank you for the comments @bigbike ! Since a leaf node can only have 0 or 1 sensors in its nodeSensorSuite, I changed the logic here a bit to just get the first sensor in the map and add/remove it.

I also call addSensorToParentNodeSensorSuite() in Sensor's constructor now.

vauduong replied:
Can you confirm if this is a valid way to do so? I know that if the map is not empty, getSensors().begin() should be valid, but I still check if it is not end() just in case.

vauduong replied:
Addressed in 1:1 with Yili

erikwijmans commented on src/esp/sensor/Sensor.cpp:
```diff
@@ -63,18 +85,28 @@ void Sensor::setTransformationFromSpec() {
+}
+
+sensor::Sensor& SensorSuite::get(const std::string& uuid) const {
+  return sensors_.at(uuid).get();
```
I forget, probably makes sense to use an `ESP_CHECK` for consistency here.

bigbike replied:
@vauduong : Here is another place you never addressed.

vauduong replied:
That's really strange, it was in the latest commit, I do not know why it did not show up but I will make sure it's there.  https://github.com/facebookresearch/habitat-sim/pull/1118/commits/09313a08d729d0032d79fa4f54b8df0ee0ce196a#diff-98b26709481602c9b0da1af7059b8153536d6bd0d1cfeb0889dec0dbbffb4d21R99-R101

vauduong commented:
@mosra @bigbike @Skylion007 @erikwijmans @aclegg3 @eundersander @jturner65 

Thank you everyone for the feedback! I've addressed all the feedback so far, and am looking forwards to a final pass on this PR. There are some comments about the python implementation that have been left for reference, but they will be addressed in a separate PR.

bigbike commented on src/esp/scene/SceneNode.cpp:
```diff
@@ -12,26 +14,171 @@ namespace scene {
+  std::map<std::string, std::reference_wrapper<sensor::Sensor>>::iterator it =
+      nodeSensorSuite_->getSensors().begin();
+  if (it != nodeSensorSuite_->getSensors().end()) {
+    CORRADE_ASSERT(static_cast<scene::SceneNode*>(this->parent()) != nullptr,
```
Here, the same thing. `static_cast` is the problem.
And you just simply copy and paste whatever you have in the `addSensorToParentNodeSensorSuite`.
**But clearly the logic is different in an "add" and a "remove" functions.**
Here, if you find you cannot dynamic_cast to its parent, one case is that the parent (SceneNode) is already deconstructed. So the conversion fails., which means in this case, you do not need to do anything. (The logic you have here is to let the program exit due to ASSERTION failure. It is totally wrong.)

vauduong replied:
Ok, thanks for the clarification!

bigbike commented on src/esp/scene/SceneNode.cpp:
```diff
@@ -12,26 +14,171 @@ namespace scene {
+    SceneNode* currentNode = dynamic_cast<SceneNode*>(this->parent());
+    while (currentNode && !SceneGraph::isRootNode(*currentNode)) {
+      currentNode->getSubtreeSensorSuite().remove(sensor.first);
+      currentNode = dynamic_cast<SceneNode*>(currentNode->parent());
```
Again, what if the currentNode becomes `nullptr`? It can happen during the deconstruction process.

vauduong replied:
We check that currentNode is not nullptr in the while loop condition.

bigbike reviewed (changes requested):
I did not even go through all the PR. Too many problems need to be addressed already.
Stop reviewing now.

vivian commented:
Hi, I've been repeatedly pinged in this thread in review comments. I believe you're meaning to ping someone else. Just letting you know.

bigbike commented:
@vivian: Sorry! I sincerely apologize! It was my mistake. I am pretty sure this would be the last time to ping you in this PR.

vauduong commented on src/esp/scene/SceneNode.cpp:
```diff
@@ -12,26 +14,171 @@ namespace scene {
+      nodeSensorSuite_->getSensors().begin();
+  if (it != nodeSensorSuite_->getSensors().end()) {
+    CORRADE_ASSERT(static_cast<scene::SceneNode*>(this->parent()) != nullptr,
+                   "SceneNode::addSensorToParentNodeSensorSuite(): Parent null "
```
Sorry this is a late-night typo, I meant Parent node, will be more careful and look thoroughly

vauduong commented on src/esp/bindings_js/bindings_js.cpp:
```diff
@@ -204,6 +203,15 @@ EMSCRIPTEN_BINDINGS(habitat_sim_bindings_js) {
       .function("setLocalTransform", &Sensor_setLocalTransform)
       .function("specification", &Sensor::specification);
 
+  em::class_<SensorSuite>("SensorSuite")
```
Took a deeper look and none of the demos except the vr demo will call getSensorSuite, but there are some issues that I posted about in slack and am still investigating. 

emscripten can't use reference types to interact with js, only pointers and primitives https://stackoverflow.com/questions/55905148/emscripten-pass-stl-c-map-parameter

It seems there's a workaround to write a policy to get pointers from references, but I'm not sure if this is what we want to do here.  https://github.com/emscripten-core/emscripten/issues/3480

Can anyone provide some guidance?

eundersander replied:
How about have JS getSensors return pointers instead of references? And then instead of binding it directly to SensorSuite::getSensors, bind it to some new C++ helper function that constructs the map of pointers.

vauduong replied:
Thank you for the suggestion eundersander! vr demo is fixed and working

bigbike commented on src/esp/scene/SceneNode.cpp:
```diff
@@ -12,26 +14,173 @@ namespace scene {
 
 SceneNode::SceneNode(SceneNode& parent)
```
Hold on here:
You have code duplication in the constructors.
You need constructor delegation.

bigbike commented on src/esp/scene/SceneNode.cpp:
```diff
@@ -12,26 +14,173 @@ namespace scene {
+  }
+}
+
+void SceneNode::removeSubtreeSensorsFromAncestors() {
```
```
if (SceneGraph::isRootNode(*this)) {
  return;
}
```

bigbike commented on src/esp/scene/SceneNode.cpp:
```diff
@@ -12,26 +14,173 @@ namespace scene {
+void SceneNode::removeSubtreeSensorsFromAncestors() {
+  for (const auto& sensor : subtreeSensorSuite_->getSensors()) {
+    SceneNode* currentNode = dynamic_cast<SceneNode*>(this->parent());
+    while (currentNode && !SceneGraph::isRootNode(*currentNode)) {
```
if `currentNode` is the root node, then this piece of code will not remove the sensors from the `currentNode`.

bigbike replied:
Go back to the slack channel, read the code I gave you very carefully, and try to understand the difference between my code and the code you wrote here.

bigbike replied:
In your tests, DO check if the number of sensors is correct in the root node!!

vauduong replied:
Thank you, this is really important. Here is the code from slack:
```
 SceneNode* currentNode = this;
  do {
    currentNode = dynamic_cast<SceneNode*>(currentNode->parent());
    // no need to worry currentNode could be nullptr, no chance
    currentNode->getSubtreeSensorSuite().getSensors().erase(sensor.first);
  } while (!SceneGraph::isRootNode(*currentNode));
  
```
We had discussed previously that currentNode could never be nullptr, which made sense because the currentNode would never be nullptr before reaching rootNode's parent, which it would not reach due to the while condition. However, this currentNode was actually coming back nullptr sometimes, and this was causing the segfault when running examples.py. My update to check that currentNode was not nullptr fixed the segfault, but overlooked adding/removing sensors from the root node's SensorSuites.

I have updated the logic to check that currentNode is not nullptr and to also update rootNode. I have also added checks for rootNode's SensorSuites in SensorTest.

Thanks and let me know if you have any questions.

bigbike commented on src/esp/scene/SceneNode.cpp:
```diff
@@ -12,26 +14,173 @@ namespace scene {
+void SceneNode::addSubtreeSensorsToAncestors() {
+  for (const auto& sensor : subtreeSensorSuite_->getSensors()) {
+    SceneNode* currentNode = dynamic_cast<SceneNode*>(this->parent());
+    while (currentNode && !SceneGraph::isRootNode(*currentNode)) {
```
So it means the root node will never ever update its sensorSuite.

bigbike commented on src/esp/scene/SceneNode.cpp:
```diff
@@ -12,26 +14,173 @@ namespace scene {
+}
+
+void SceneNode::addSensorToParentNodeSensorSuite() {
+  // Only add a sensor if this is a leaf node and sensor exists
```
if this is called from the root node, and somehow the root node is marked as a leaf node, this function should just return.

bigbike commented on src/esp/scene/SceneNode.cpp:
```diff
@@ -12,26 +14,173 @@ namespace scene {
+}
+
+void SceneNode::removeSensorFromParentNodeSensorSuite() {
+  // Only remove a sensor if this is a leaf node and sensor exists
```
again, what if `this` is the root node of the scene graph, and it is a leaf?

vauduong commented on src/esp/scene/SceneNode.cpp:
```diff
@@ -12,26 +14,173 @@ namespace scene {
+    return;
+  }
+  // There only exists 0 or 1 sensors in a leaf nodeSensorSuite
+  CORRADE_INTERNAL_ASSERT(nodeSensorSuite_->getSensors().size() == 1);
```
Yes, we have this check already in the line above when we check if nodeSensorSuite is empty:
```
// Only add a sensor if this is a leaf node and sensor exists
  if (!(getSceneNodeTags() & SceneNodeTag::Leaf) ||
      nodeSensorSuite_->getSensors().empty()) {
    return;
  }
```

bigbike replied:
I see. Thank you.

eundersander commented on src/esp/agent/Agent.h:
```diff
@@ -133,14 +133,24 @@ class Agent : public Magnum::SceneGraph::AbstractFeature3D {
+   * this Agent's SceneNode and its children values.
+   * NOTE: This is only called in JS, as emscripten needs pointers
+   */
+  std::map<std::string, sensor::Sensor::ptr> jsGetSubtreeSensors() {
```
Let's avoid enlarging this class's public interface unnecessarily. I had in mind a global static function defined in bindings_js.cpp and not exposed to other parts of the codebase. We don't want to litter the C++ codebase with js-specific code.

eundersander commented on src/esp/bindings_js/bindings_js.cpp:
```diff
@@ -276,7 +261,7 @@ EMSCRIPTEN_BINDINGS(habitat_sim_bindings_js) {
       .function("hasAction", &Agent::hasAction)
       .function("act", &Agent::act)
-      .function("getSubtreeSensors", &Agent::getSubtreeSensors);
+      .function("getSubtreeSensors", &Agent::jsGetSubtreeSensors);
```
You can bind to a free function (a static global function defined in this file) instead of a Agent member method.

vauduong replied:
Thanks for the clarification Eric! That's been addressed and is tested and working.

bigbike commented on src/esp/scene/SceneNode.cpp:
```diff
@@ -3,35 +3,197 @@
+SceneNode::SceneNode() : Mn::SceneGraph::AbstractFeature3D{*this} {
   setCachedTransformations(Mn::SceneGraph::CachedTransformation::Absolute);
   absoluteTransformation_ = absoluteTransformation();
+  // nodeSensorSuite_ and subtreeSensorSuite_ will be released upon SceneNode's
```
`// Once created, nodeSensorSuite_ and subtreeSensorSuite_ are features owned by the scene graph node.
No need to release them in the destructor since magnum scene graph will handle it. `

vauduong replied:
Thank you! Let me know if there are any other changes you think I should make.

bigbike commented on src/esp/scene/SceneNode.cpp:
```diff
@@ -12,26 +14,173 @@ namespace scene {
+}
+
+SceneNode& SceneNode::createChild(SceneNodeTags childNodeTags) {
+  CORRADE_ASSERT(
```
If you already have `ESP_CHECK`, you do not need the `CORRADE_ASSERT` any more. You should read the comments about the `ESP_CHECK` before you use it.

bigbike commented on src/esp/scene/SceneNode.cpp:
```diff
@@ -12,26 +14,173 @@ namespace scene {
+  // Perform same internal checks as magnum to ensure newParent is a valid new
+  // parent before updating any SensorSuites
+  // Parent can not be leaf node
+  CORRADE_ASSERT(!(newParent->getSceneNodeTags() & SceneNodeTag::Leaf),
```
Again, `CORRADE_ASSERT` is not needed any more.

bigbike commented on src/esp/scene/SceneNode.cpp:
```diff
@@ -3,35 +3,191 @@
-    : Mn::SceneGraph::AbstractFeature3D{*this} {
-  setParent(&parent);
-  setId(parent.getId());
+SceneNode::SceneNode() : Mn::SceneGraph::AbstractFeature3D{*this} {
```
Wait. Just realize this is a new constructor. Is it a must have?
The consequence is that with this default constructor, the user can now create a scene node without connecting it to any  scene graph. (dangling scene node).

bigbike replied:
OK. I see the code below. It is not a public constructor. But again, is it a must have?

bigbike commented on src/esp/scene/SceneNode.cpp:
```diff
@@ -3,35 +3,191 @@
-  setParent(&parentNode);
-  setCachedTransformations(Mn::SceneGraph::CachedTransformation::Absolute);
-  absoluteTransformation_ = absoluteTransformation();
+SceneNode::SceneNode(SceneNode& parent) : SceneNode() {
```
If `SceneNode()` constructor is removed, can this one call the delegating constructor `SceneNode::SceneNode(MagnumScene& parentNode)` below?
 You may need a type conversion in the argument.

bigbike commented on src/esp/sensor/SensorFactory.h:
```diff
@@ -6,16 +10,33 @@ namespace sensor {
-  static sensor::SensorSuite createSensors(
-      scene::SceneNode& node,
-      const sensor::SensorSetup& sensorSetup);
+  static void deleteSensor(scene::SceneNode& node, const std::string& uuid);
```
Just read the comments. So consider call it `deleteSubtreeSensor` to avoid confusion.
